data-prepper-plugins/service-map-stateful/src/main/java/com/amazon/dataprepper/plugins/prepper/ServiceMapStatefulPrepper.java

```diff
@@ -165,88 +165,81 @@ private static File createPath(File path) {
      * @return Set of Record<String> containing json representation of ServiceMapRelationships found
      */
     private Collection<Record<String>> evaluateEdges() {
+        LOG.info("Evaluating service map edges");
         try {
-            final Stream<ServiceMapRelationship> previousStream = previousWindow.iterate(relationshipIterationFunction, preppersCreated.get(), thisPrepperId).stream().flatMap(serviceMapEdgeStream -> serviceMapEdgeStream);
-            final Stream<ServiceMapRelationship> currentStream = currentWindow.iterate(relationshipIterationFunction, preppersCreated.get(), thisPrepperId).stream().flatMap(serviceMapEdgeStream -> serviceMapEdgeStream);
-
-            final Collection<Record<String>> serviceDependencyRecords =
-                    Stream.concat(previousStream, currentStream).filter(Objects::nonNull)
-                            .filter(serviceMapRelationship -> !relationshipState.contains(serviceMapRelationship))
-                            .map(serviceMapRelationship -> {
-                                try {
-                                    relationshipState.add(serviceMapRelationship);
-                                    return new Record<>(OBJECT_MAPPER.writeValueAsString(serviceMapRelationship));
-                                } catch (JsonProcessingException e) {
-                                    throw new RuntimeException(e);
-                                }
-                            })
-                            .collect(Collectors.toSet());
-
-            if (edgeEvaluationLatch == null) {
-                initEdgeEvaluationLatch();
-            }
-            doneEvaluatingEdges();
-            waitForEvaluationFinish();
+            final Collection<Record<String>> serviceDependencyRecords = new HashSet<>();
+
+            serviceDependencyRecords.addAll(iteratePrepperState(previousWindow));
+            serviceDependencyRecords.addAll(iteratePrepperState(currentWindow));
+            LOG.info("Done evaluating service map edges");
+
+            // Wait for all workers before rotating windows
+            allThreadsCyclicBarrier.await();
 
             if (isMasterInstance()) {
                 rotateWindows();
```

Comment:
Is it possible to insert rotateWindows into the barrierAction? 
```
public CyclicBarrier(int parties, Runnable barrierAction)
```
Then we might get rid of the masterInstance and also avoid calling await twice

Reply:
I guess that might need removing the @SingleThread concurrency protocol as prerequisite. Feel free to push back to separate PR if feasible.

Reply:
Yep, will eventually phase out the concept of `masterInstance` and remove the `@SingleThread` annotation as you mentioned (in a separate PR though).

Reply:
> Is it possible to insert rotateWindows into the barrierAction?

Interesting, hadn't see that 👍. Should be possible, though would have to research a bit more to confirm the order of events is what we expect.